File: scripts/insert_test_data.py
#!/usr/bin/env python3
"""
插入300万条测试数据到日志服务
服务名称: zhenhaotou
包含字段: adv_id, aweme_id, plan_id (随机生成)
"""

import grpc
import random
import string
import time
import threading
from concurrent.futures import ThreadPoolExecutor
from datetime import datetime, timedelta
import sys
import os

# 添加proto文件路径
sys.path.append(os.path.join(os.path.dirname(__file__), '..', 'clients', 'python'))

try:
    import log_service_pb2
    import log_service_pb2_grpc
except ImportError:
    print("错误: 无法导入protobuf文件")
    print("请确保已经生成了Python的protobuf文件")
    print("运行: cd clients/python && ./setup_and_run.sh")
    sys.exit(1)

# 配置常量
TOTAL_RECORDS = 3000000  # 300万条记录
BATCH_SIZE = 1000        # 每批1000条
MAX_WORKERS = 10         # 最大并发数
SERVICE_NAME = "zhenhaotou"
GRPC_ADDRESS = "localhost:50051"

# 日志级别
LOG_LEVELS = [
    log_service_pb2.DEBUG,
    log_service_pb2.INFO,
    log_service_pb2.WARN,
    log_service_pb2.ERROR,
    log_service_pb2.FATAL
]

# 日志消息模板
LOG_MESSAGES = [
    "用户访问页面",
    "API请求处理", 
    "数据库查询执行",
    "缓存更新操作",
    "文件上传完成",
    "用户登录成功",
    "订单创建完成",
    "支付处理成功",
    "数据同步完成",
    "任务执行结束"
]

class DataInserter:
    def __init__(self):
        self.channel = None
        self.stub = None
        self.total_inserted = 0
        self.start_time = None
        self.lock = threading.Lock()
        
    def connect(self):
        """连接到gRPC服务"""
        print(f"连接到gRPC服务器: {GRPC_ADDRESS}")
        self.channel = grpc.insecure_channel(GRPC_ADDRESS)
        self.stub = log_service_pb2_grpc.LogServiceStub(self.channel)
        
        # 测试连接
        try:
            response = self.stub.WriteLog(log_service_pb2.WriteLogRequest(
                log_entry=self.generate_log_entry()
            ))
            if response.success:
                print(f"连接测试成功，日志ID: {response.log_id}")
            else:
                raise Exception(f"测试写入失败: {response.error_message}")
        except Exception as e:
            print(f"连接测试失败: {e}")
            raise
    
    def generate_random_id(self, prefix):
        """生成随机ID"""
        timestamp = str(int(time.time() * 1000000))
        random_suffix = ''.join(random.choices(string.digits, k=6))
        return f"{prefix}_{timestamp}_{random_suffix}"
    
    def generate_log_entry(self):
        """生成随机日志条目"""
        # 生成随机时间戳（最近30天内）
        now = datetime.now()
        random_days = random.randint(0, 29)
        random_hours = random.randint(0, 23)
        random_minutes = random.randint(0, 59)
        random_seconds = random.randint(0, 59)
        
        timestamp = now - timedelta(
            days=random_days,
            hours=random_hours, 
            minutes=random_minutes,
            seconds=random_seconds
        )
        
        return log_service_pb2.LogEntry(
            service_name=SERVICE_NAME,
            level=random.choice(LOG_LEVELS),
            message=f"{random.choice(LOG_MESSAGES)} - {random.randint(1, 10000)}",
            timestamp=timestamp.isoformat(),
            metadata={
                "adv_id": self.generate_random_id("adv"),
                "aweme_id": self.generate_random_id("aweme"), 
                "plan_id": self.generate_random_id("plan"),
                "user_id": str(random.randint(1, 100000)),
                "region": random.choice(["北京", "上海", "广州", "深圳", "杭州"]),
                "platform": random.choice(["iOS", "Android", "Web", "Desktop"])
            },
            trace_id=self.generate_random_id("trace"),
            span_id=self.generate_random_id("span")
        )
    
    def insert_batch(self, batch_num, batch_size):
        """插入一批数据"""
        try:
            # 生成批次数据
            log_entries = []
            for _ in range(batch_size):
                log_entries.append(self.generate_log_entry())
            
            # 执行批量插入
            response = self.stub.BatchWriteLog(log_service_pb2.BatchWriteLogRequest(
                log_entries=log_entries
            ))
            
            if response.success:
                with self.lock:
                    self.total_inserted += len(log_entries)
                    if batch_num % 100 == 0:
                        elapsed = time.time() - self.start_time
                        rate = self.total_inserted / elapsed
                        print(f"批次 {batch_num}: 已插入 {self.total_inserted:,} 条记录, "
                              f"速度: {rate:.0f} 条/秒")
                return True
            else:
                print(f"批次 {batch_num} 失败: {response.error_message}")
                return False
                
        except Exception as e:
            print(f"批次 {batch_num} 异常: {e}")
            return False
    
    def run_insertion(self):
        """运行数据插入"""
        print(f"开始插入测试数据...")
        print(f"配置: 总记录数={TOTAL_RECORDS:,}, 批次大小={BATCH_SIZE}, 并发数={MAX_WORKERS}")
        
        self.start_time = time.time()
        
        # 计算批次数
        total_batches = (TOTAL_RECORDS + BATCH_SIZE - 1) // BATCH_SIZE
        print(f"将执行 {total_batches} 个批次")
        
        # 使用线程池执行批量插入
        with ThreadPoolExecutor(max_workers=MAX_WORKERS) as executor:
            futures = []
            
            for batch_num in range(total_batches):
                # 计算当前批次大小
                remaining = TOTAL_RECORDS - batch_num * BATCH_SIZE
                current_batch_size = min(BATCH_SIZE, remaining)
                
                future = executor.submit(self.insert_batch, batch_num + 1, current_batch_size)
                futures.append(future)
            
            # 等待所有任务完成
            success_count = 0
            for future in futures:
                if future.result():
                    success_count += 1
        
        # 输出统计信息
        duration = time.time() - self.start_time
        print("\n" + "="*50)
        print("数据插入完成！")
        print(f"成功批次: {success_count}/{total_batches}")
        print(f"插入记录: {self.total_inserted:,}/{TOTAL_RECORDS:,}")
        print(f"总耗时: {duration:.2f} 秒")
        print(f"插入速度: {self.total_inserted/duration:.0f} 条/秒")
        print("="*50)
    
    def close(self):
        """关闭连接"""
        if self.channel:
            self.channel.close()

def main():
    inserter = DataInserter()
    
    try:
        inserter.connect()
        inserter.run_insertion()
    except KeyboardInterrupt:
        print("\n中断插入操作...")
    except Exception as e:
        print(f"插入失败: {e}")
        sys.exit(1)
    finally:
        inserter.close()
    
    print("\n可以通过以下方式验证数据:")
    print("1. 查看管理界面: http://localhost:3000")
    print("2. 使用客户端查询: python clients/python/client.py")
    print("3. 直接查询MongoDB")

if __name__ == "__main__":
    main()

File: scripts/insert_test_data.go
package main

import (
	"context"
	"fmt"
	"log"
	"math/rand"
	"strconv"
	"sync"
	"time"

	pb "log-service/proto"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

const (
	// 总共要插入的数据量
	TOTAL_RECORDS = 3000000
	// 每批插入的数量 (减少批次大小)
	BATCH_SIZE = 500
	// 并发协程数 (减少并发数)
	CONCURRENT_WORKERS = 5
	// 服务名
	SERVICE_NAME = "zhenhaotou"
	// gRPC服务地址
	GRPC_ADDRESS = "localhost:50051"
)

// 日志级别列表
var logLevels = []pb.LogLevel{
	pb.LogLevel_DEBUG,
	pb.LogLevel_INFO,
	pb.LogLevel_WARN,
	pb.LogLevel_ERROR,
	pb.LogLevel_FATAL,
}

// 日志消息模板
var logMessages = []string{
	"用户访问页面",
	"API请求处理",
	"数据库查询执行",
	"缓存更新操作",
	"文件上传完成",
	"用户登录成功",
	"订单创建完成",
	"支付处理成功",
	"数据同步完成",
	"任务执行结束",
}

// 生成随机ID
func generateRandomID(prefix string) string {
	return fmt.Sprintf("%s_%d_%d", prefix, time.Now().UnixNano(), rand.Intn(999999))
}

// 生成随机日志条目
func generateLogEntry() *pb.LogEntry {
	now := time.Now()

	// 随机生成时间戳（最近30天内）
	randomDays := rand.Intn(30)
	randomHours := rand.Intn(24)
	randomMinutes := rand.Intn(60)
	randomSeconds := rand.Intn(60)

	timestamp := now.AddDate(0, 0, -randomDays).
		Add(-time.Duration(randomHours) * time.Hour).
		Add(-time.Duration(randomMinutes) * time.Minute).
		Add(-time.Duration(randomSeconds) * time.Second)

	return &pb.LogEntry{
		ServiceName: SERVICE_NAME,
		Level:       logLevels[rand.Intn(len(logLevels))],
		Message:     logMessages[rand.Intn(len(logMessages))] + " - " + strconv.Itoa(rand.Intn(10000)),
		Timestamp:   timestamp.Format(time.RFC3339),
		Metadata: map[string]string{
			"adv_id":   generateRandomID("adv"),
			"aweme_id": generateRandomID("aweme"),
			"plan_id":  generateRandomID("plan"),
			"user_id":  strconv.Itoa(rand.Intn(100000)),
			"region":   []string{"北京", "上海", "广州", "深圳", "杭州"}[rand.Intn(5)],
			"platform": []string{"iOS", "Android", "Web", "Desktop"}[rand.Intn(4)],
		},
		TraceId: generateRandomID("trace"),
		SpanId:  generateRandomID("span"),
	}
}

// 批量插入工作函数
func batchInsertWorker(client pb.LogServiceClient, batchChan <-chan []*pb.LogEntry, wg *sync.WaitGroup, workerID int) {
	defer wg.Done()

	batchCount := 0
	for batch := range batchChan {
		// 执行批量插入
		resp, err := client.BatchWriteLog(context.Background(), &pb.BatchWriteLogRequest{
			LogEntries: batch,
		})

		batchCount++
		if err != nil {
			log.Printf("Worker %d - Batch %d failed: %v", workerID, batchCount, err)
			continue
		}

		if !resp.Success {
			log.Printf("Worker %d - Batch %d failed: %s", workerID, batchCount, resp.ErrorMessage)
			continue
		}

		log.Printf("Worker %d - Batch %d completed, inserted %d records", workerID, batchCount, len(batch))
	}

	log.Printf("Worker %d completed, processed %d batches", workerID, batchCount)
}

// 生成批次数据
func generateBatches(batchChan chan<- []*pb.LogEntry) {
	defer close(batchChan)

	totalBatches := (TOTAL_RECORDS + BATCH_SIZE - 1) / BATCH_SIZE
	log.Printf("将生成 %d 个批次，每批次 %d 条记录", totalBatches, BATCH_SIZE)

	for batchNum := 0; batchNum < totalBatches; batchNum++ {
		// 计算当前批次应该生成多少条记录
		remainingRecords := TOTAL_RECORDS - batchNum*BATCH_SIZE
		currentBatchSize := BATCH_SIZE
		if remainingRecords < BATCH_SIZE {
			currentBatchSize = remainingRecords
		}

		// 生成当前批次的日志条目
		batch := make([]*pb.LogEntry, currentBatchSize)
		for i := 0; i < currentBatchSize; i++ {
			batch[i] = generateLogEntry()
		}

		batchChan <- batch

		if (batchNum+1)%100 == 0 {
			log.Printf("已生成 %d/%d 批次", batchNum+1, totalBatches)
		}
	}

	log.Printf("所有批次生成完成，总计 %d 批次", totalBatches)
}

func main() {
	log.Printf("开始插入测试数据...")
	log.Printf("配置: 总记录数=%d, 批次大小=%d, 并发数=%d", TOTAL_RECORDS, BATCH_SIZE, CONCURRENT_WORKERS)

	startTime := time.Now()

	// 连接到gRPC服务器
	log.Printf("连接到gRPC服务器: %s", GRPC_ADDRESS)
	conn, err := grpc.NewClient(GRPC_ADDRESS, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		log.Fatalf("连接服务器失败: %v", err)
	}
	defer conn.Close()

	// 创建客户端
	client := pb.NewLogServiceClient(conn)

	// 测试连接
	log.Printf("测试连接...")
	testResp, err := client.WriteLog(context.Background(), &pb.WriteLogRequest{
		LogEntry: generateLogEntry(),
	})
	if err != nil {
		log.Fatalf("测试连接失败: %v", err)
	}
	if !testResp.Success {
		log.Fatalf("测试写入失败: %s", testResp.ErrorMessage)
	}
	log.Printf("连接测试成功，日志ID: %s", testResp.LogId)

	// 创建批次通道
	batchChan := make(chan []*pb.LogEntry, CONCURRENT_WORKERS*2)

	// 启动工作协程
	var wg sync.WaitGroup
	for i := 0; i < CONCURRENT_WORKERS; i++ {
		wg.Add(1)
		go batchInsertWorker(client, batchChan, &wg, i+1)
	}

	// 生成并发送批次数据
	log.Printf("开始生成批次数据...")
	generateBatches(batchChan)

	// 等待所有工作协程完成
	log.Printf("等待所有插入操作完成...")
	wg.Wait()

	duration := time.Since(startTime)
	log.Printf("数据插入完成！")
	log.Printf("总耗时: %v", duration)
	log.Printf("插入速度: %.2f 条/秒", float64(TOTAL_RECORDS)/duration.Seconds())
	log.Printf("平均延迟: %.2f ms/批次", float64(duration.Milliseconds())/float64((TOTAL_RECORDS+BATCH_SIZE-1)/BATCH_SIZE))
}
